[RFC] cl_exp_tensor (#1006)
KhronosGroup/OpenCL-Docs: extensions/cl_exp_tensor.asciidoc

:data-uri:
:icons: font
//include::../config/attribs.txt[]
//include::{generated}/api/api-dictionary.asciidoc[]
:source-highlighter: coderay

= cl_exp_tensor

This extension provides new buffer abstraction - tensor objects - for
managing N-dimensional data.

== XXX - Not complete yet!!!

== Name Strings

`cl_exp_tensor`

== Contact

TODO

== Contributors

Henry Linjamäki, Intel. +
Pekka Jääslkeläinen, Intel and Tampere University. +
Ben Ashbaugh, Intel. +

== Notice

TODO

== Status

Draft spec, NOT APPROVED!!

== Version

Built On: {docdate} +
Version: 0.2.0

== Dependencies

This extension is written against the OpenCL Specification version 3.0.14.

This extension requires OpenCL 1.2 or later.

== Overview

The extension provides new tensor object abstraction. Tensor objects

> are similar to image types in regard they represents N-dimensional

Reviewer: ```suggestion
are similar to image types in regard that they represent N-dimensional
```

> data of some application chosen data type and they may be mapped to

Reviewer: ```suggestion
data of an application chosen data type and they may be mapped to
```

> dedicated hardware except that

Reviewer: ```suggestion
dedicated hardware, with the following key differences:
```

> * higher than 3-dimensional data can be supported (limited by

Reviewer: ```suggestion
* Higher than 3-dimensional data can be supported (limited by
```

  devices' capabilities).

> * applications may choose how the data elements of the tensors are

Reviewer: ```suggestion
* Applications may choose how the data elements of the tensors are
```

  laid out in the buffers using the tensor layout descriptions
  provided in this extension.

> Applications may also choose the memory layouts of the tensors be

Reviewer: ```suggestion
Applications may also choose the memory layouts of the tensors to be
```

implementation-specified, letting the driver to optimize the tensor
data layout for better performance or to lay out the data as required by

> hardware functions (e.g. exposed via builtin kernels).

Reviewer: ```suggestion
hardware accelerated functions (e.g. exposed via builtin kernels).
```

> The scope of this extension to provide host APIs for creating tensor

Reviewer: ```suggestion
The scope of this extension is to provide host APIs for creating tensor
```

objects and transfer data between tensors, host and other memory
objects.

A separate extension implemented on top of this extension,

> cl_exp_defined_builtin_kernels which provides "defined built-in

Reviewer: ```suggestion
cl_exp_defined_builtin_kernels provides "defined built-in
```

> kernels" (DKBs) which operates on tensors. It also provides mechanism

Reviewer: ```suggestion
kernels" (DKBs) which can operate on tensors. It also provides mechanism
```

for drivers to create DBKs that are optimized for the tensor arguments
they operate on.

> 

Reviewer: Perhaps mention that it's designed for use with command buffers for whole task-graph optimizations. And maybe also about Vulkan layering (if those referred APIs are public).

  cl_mutable_command_khr* mutable_handle);
----

== New API Types

[source,c]
----
typedef cl_uint cl_tensor_layout_type_exp;
typedef cl_uint cl_tensor_dim_exp;
typedef cl_uint cl_tensor_layout_ml_type_exp;
typedef cl_properties cl_tensor_properties_exp;

#define CL_TENSOR_DESC_MAX_RANK_EXP       20u
#define CL_TENSOR_DESC_MAX_PROPERTIES_EXP 16u

typedef struct cl_tensor_desc_exp {
    cl_uint               rank;
    cl_tensor_datatype    dtype;
    cl_tensor_properties_exp  properties[CL_TENSOR_DESC_MAX_PROPERTIES_EXP]
    cl_tensor_shape       shape[CL_TENSOR_DESC_MAX_RANK_EXP];
    const void*           layout;
    cl_tensor_layout_type_exp layout_type;
} cl_tensor_desc_exp;

typedef struct cl_tensor_layout_blas_exp {
    cl_tensor_dim_exp    leading_dims[CL_TENSOR_DESC_MAX_RANK_EXP];
} cl_tensor_layout_blas_exp;

typedef struct cl_tensor_layout_blas_pitched_exp {
    cl_tensor_dim_exp    leading_dims[CL_TENSOR_DESC_MAX_RANK_EXP];
    cl_tensor_stride     leading_strides[CL_TENSOR_DESC_MAX_RANK_EXP];
} cl_tensor_layout_blas_pitched__exp;

typedef struct cl_tensor_layout_ml_exp {
  cl_tensor_layout_ml_type_exp ml_type;
} cl_tensor_layout_ml_exp;
----

== New API Enums

> Accepted value for _properties_ parameter to

Reviewer: ```suggestion
Accepted value for the _properties_ parameter to
```

* CL_INVALID_VALUE if `origin[i] + region[i] > tensor_shape[i]` at any
  dimension `i` in range `[0, tensor_rank)`.

* CL_INVALID_EVENT_WAIT_LIST if _event_wait_list_ is NULL and
  _num_events_in_wait_list_ > 0, or _event_wait_list_ is not NULL and
  _num_events_in_wait_list_ is 0, or if event objects in
  _event_wait_list_ are not valid events.

* CL_EXEC_STATUS_ERROR_FOR_EVENTS_IN_WAIT_LIST if the read and write
  operations are blocking and the execution status of any of the
  events in _event_wait_list_ is a negative integer value.

* CL_MEM_OBJECT_ALLOCATION_FAILURE if there is a failure to allocate
  memory for data store associated with memory object the _tensor_ is
  bound to.

* CL_OUT_OF_RESOURCES if there is a failure to allocate resources
  required by the OpenCL implementation on the device.

* CL_OUT_OF_HOST_MEMORY if there is a failure to allocate resources
  required by the OpenCL implementation on the host.

// TODO: add clEnqueueFillTensor?

--
// End (Add the following to Section 5.2.2, *Reading, Writing and Copying Buffer Objects*)

(Add the following to Section 5.17.5, *Recording Commands to a Command-Buffer*) ::
+
--

> If *cl_khr_command_buffer* is supported, then the following command

Reviewer: Refer to this extension in the intro spec, state that optionally cl_khr_command_buffer is extended.

  point to.

[[tensor-dtypes-table]]
.Tensor element types. The API type indicates the corresponding type for copying elements from an host allocation / buffer object to tensor or vice versa.
[cols="1,1,1",stripes=even]
|===
| *Tensor element data type* | *Description* | *API type*

| CL_TENSOR_DTYPE_BOOL | Data type representing true or false.  |
cl_uchar. footnote:[zero and non-zero bytes are interpreted as false
and true values, respectively.]

| CL_TENSOR_DTYPE_INT4_EXP        | 4-bit signed integer.            | cl_char.
| CL_TENSOR_DTYPE_INT8_EXP        | 8-bit signed integer.            | cl_char.
| CL_TENSOR_DTYPE_INT16_EXP       | 16-bit signed integer.           | cl_short.
| CL_TENSOR_DTYPE_INT32_EXP       | 32-bit signed integer.           | cl_int.
| CL_TENSOR_DTYPE_INT64_EXP       | 64-bit signed integer.           | cl_long.
| CL_TENSOR_DTYPE_UINT8_EXP       | 8-bit unsigned integer.          | cl_uchar.
| CL_TENSOR_DTYPE_UINT16_EXP      | 16-bit unsigned integer.         | cl_ushort.
| CL_TENSOR_DTYPE_UINT32_EXP      | 32-bit unsigned integer.         | cl_uint.
| CL_TENSOR_DTYPE_UINT64_EXP      | 64-bit unsigned integer.         | cl_ulong.

> | CL_TENSOR_DTYPE_FP8_EXP         | Half precision floating-point.   | cl_char.

Reviewer: Refer to a specific FP8 format.

| CL_TENSOR_DTYPE_FP16_EXP        | Half precision floating-point.   | cl_half.
| CL_TENSOR_DTYPE_BFLOAT16_EXP    | 16-bit brain floating-point.     | cl_ushort
| CL_TENSOR_DTYPE_FP32_EXP        | Single precision floating-point. | cl_float.
| CL_TENSOR_DTYPE_FP64_EXP        | Double precision floating-point. | cl_double.
| CL_TENSOR_DTYPE_COMPLEX64_EXP   | 64-bit complex floating-point with
  32-bit real and imaginary part. | cl_float2
| CL_TENSOR_DTYPE_COMPLEX128_EXP  | 128-bit complex floating-point with
  64-bit real and imaginary part. | cl_double2
|===

[[layout-types-table]]
.Optional tensor memory layout types.
[cols="1,1,4",stripes=even]
|===
| *layout type* | *tensor layout type* | *Description*

> | CL_TENSOR_LAYOUT_OPAQUE_EXP | N/A | The tensor don't have application

Reviewer: ```suggestion
| CL_TENSOR_LAYOUT_OPAQUE_EXP | N/A | The tensor doesn't have application
```

  defined memory layout. Driver controls the tensors layout. To read

>   or write elements of the tensor

Reviewer: Is this sentence unfinished?

| CL_TENSOR_LAYOUT_BLAS_EXP
|<<cl-tensor-layout-blas,cl_tensor_layout_blas_exp>>

> | A type that describe packed memory layout similar ones used in BLAS APIs.

Reviewer: ```suggestion
| A type that describes a packed memory layout similar ones used in BLAS APIs.
```

Reviewer: Is this accurate enough description. Is the BLAS layout well-specified somewhere to which we can refer?

> | CL_TENSOR_LAYOUT_BLAS_EXP

Reviewer: Pitched? Unfinished copy-pasta?

|<<cl-tensor-layout-blas,cl_tensor_layout_blas_pitched_exp>>
| A type that describe memory layout similar ones used in BLAS APIs.

| CL_TENSOR_LAYOUT_ML_EXP       | <<cl-tensor-layout-ml,cl_tensor_layout_ml_exp>> |

> A convenience layout type over `CL_TENSOR_LAYOUT_BLAS_EXP`.

Reviewer: What does it mean?

|===

--
// End (Add the following to new Section 5.X.Y, *Tensor Descriptions*)

[[cl-tensor-layout-blas]]
(Add the following to new Section 5.X.Y.1, *BLAS Tensor Layout*) ::
+
--

> The following structures describes packed / pitched BLAS-like memory

Reviewer: ```suggestion
The following structures describe packed / pitched BLAS-like memory
```

layout for the tensor:

[source,c]
----
typedef struct cl_tensor_layout_blas_exp {
    cl_tensor_dim_exp    leading_dims[CL_TENSOR_DESC_MAX_RANK_EXP];
} cl_tensor_layout_blas_exp;

typedef struct cl_tensor_layout_blas_pitched_exp {
    cl_tensor_dim_exp    leading_dims[CL_TENSOR_DESC_MAX_RANK_EXP];
    cl_tensor_pitch      leading_pitches[CL_TENSOR_DESC_MAX_RANK_EXP];
} cl_tensor_layout_blas_pitched_exp;

typedef struct cl_tensor_layout_ml_exp {
    cl_tensor_layout_ml_type_exp ml_type;
} cl_tensor_layout_ml_exp;
----

* _leading_dims_ describes which elements along the tensor dimension

>   are laid out in the memory. `leading_dims[0]` point to dimension

Reviewer: ```suggestion
  are laid out in the memory. `leading_dims[0]` points to the dimension
```

  whose elements are laid out first, followed by elements along

>   dimension by `leading_dims[1]` and so on. The first N elements must

Reviewer: ```suggestion
  the dimension by `leading_dims[1]` and so on. The first N elements must
```

>   be non-zero where N is tensor's rank and the values must be unique

Reviewer: ```suggestion
  be non-zero where N is a tensor's rank and the values must be unique
```

  and within range `[0, tensor_rank)`.

> * _leading_pitches_ describes distance between from an element to the

Reviewer: ```suggestion
* _leading_pitches_ describes the distance between an element to the
```

  next one for the leading dimensions in _leading_dims_. The distance
  is measured in number of elements. The first N elements must be
  non-zero where the N is tensor's rank minus one. The values of the
  array must be non-zero for the first tensor rank minus one elements
  and following conditions must hold:

** `leading_pitches[0] >= tensor_shape[leading_dims[0]]` if the tensor
   rank is greater than one and

** `leading_pitches[i + 1] >= tensor_shape[leading_dims[i]] *
  leading_pitches[i]` for `i` in `[0, tensor_rank - 1)` if the tensor
  rank is greater than two.

// ^ This condition is meant to ensure that the tensor elements at different
// coordinates don't alias.

> * _ml_type_ defines memory layout via enumerators which corresponds to

Reviewer: ```suggestion
* _ml_type_ defines the memory layout via enumerators which corresponds to
```

  predefined configurations of `cl_tensor_layout_blas_exp` structure
  as listed in <<tensor-layout-ml-types,ML tensor layout type>> table.

The memory layout descriptions map tensor coordinates to buffer's

> memory byte locations respect to buffer's base address as followed in

Reviewer: ```suggestion
memory byte locations respect to buffer's base address as in the following pseudo code example.
```

pseudo C:

[source,c]
----
size_t index = 0;
for (unsigned i = 0; i < tensor_rank - 1; i++)
  index += tensor_coordinates[leading_dims[i]] * pitches[i];
buffer_offset = index * tensor_element_size;
----

Where `pitches[i]` equals to:

* _leading_pitches_[i] for `cl_tensor_layout_blas_pitched_exp`.

* `tensor_shape[leading_dims[i]] *
  tensor_shape[leading_dims[i-1]] * ... *
  tensor_shape[leading_dims[0]]` for `cl_tensor_layout_blas_exp`.

[[tensor-layout-ml-type]]
.ML tensor layout types and their corresponding cl_tensor_layout_blas_exp configuration.
[cols="1,2",stripes=even]
|===
| *ML layout type* | *Equivalent _leading_dims_ configuration*

|CL_TENSOR_LAYOUT_ML_C_EXP    | `{}`
|CL_TENSOR_LAYOUT_ML_NC_EXP   | `{1}`
|CL_TENSOR_LAYOUT_ML_CN_EXP   | `{0}`
|CL_TENSOR_LAYOUT_ML_HW_EXP   | `{1}`
|CL_TENSOR_LAYOUT_ML_CHW_EXP  | `{2, 1}`
|CL_TENSOR_LAYOUT_ML_NCHW_EXP | `{3, 2, 1}`
|CL_TENSOR_LAYOUT_ML_NHWC_EXP | `{1, 3, 2}`
|===
--

== Sample Codes

An example usage of tensors:

[source,cpp]
----
constexpr size_t b = 64, m = 100, n = 200, k = 50;

std::vector<float> in0_data = ...;
std::vector<float> in1_data = ...;
std::vector<float> out_data(b * m * n);

> // Create tensor with opaque layout.

Reviewer: ```suggestion
// Create a tensor with an opaque layout.
```

cl_tensor_desc_exp in0_desc;
in0_desc.rank = 3;
in0_desc.properties[0] = 0;
in0_desc.shape[0] = b;
in0_desc.shape[1] = m;
in0_desc.shape[2] = k;
in0_desc.layout = nullptr;
in0_desc.layout_type = CL_TENSOR_LAYOUT_OPAQUE_EXP;

cl_int err;
cl_mem in0_tensor = clCreateBufferWithProperties(
  ctx, {CL_MEM_TENSOR_EXP, in0_desc, 0},
  CL_MEM_READ_ONLY, 0, nullptr, &err);

> // Create tensor from a host allocation using an application defined

Reviewer: ```suggestion
// Create tensor from a host allocation using an application-defined
```

// layout description for mapping elements to the tensor.
cl_tensor_desc_exp in1_desc;
in1_desc.rank = 3;
in1_desc.properties[0] = 0;
in1_desc.shape[0] = b;
in1_desc.shape[1] = k;
in1_desc.shape[2] = n;

cl_tensor_layout_blas_exp col_major;
col_major.leading_dims[0] = 1,
col_major.leading_dims[1] = 2,
in1_desc.layout = &col_major;
in1_desc.layout_type = CL_TENSOR_LAYOUT_BLAS_EXP;

cl_mem in1_tensor = clCreateBufferWithProperties(
  ctx, {CL_MEM_TENSOR_EXP, in1_desc, 0},
  CL_MEM_USE_HOST_PTR | CL_MEM_READ_ONLY, 0, in1_data.data(), &err);

> // Create another tensor with application defined layout.

Reviewer: ```suggestion
// Create another tensor with an application-defined layout.
```

cl_tensor_desc_exp out_desc;
out_desc.rank = 3;
out_desc.properties[0] = 0;
out_desc.shape[0] = b;
out_desc.shape[1] = m;
out_desc.shape[2] = n;

cl_tensor_layout_blas_exp row_major;
row_major.leading_dims[0] = 2,
row_major.leading_dims[1] = 1,
out_desc.layout = &row_major;
out_desc.layout_type = CL_TENSOR_LAYOUT_BLAS_EXP;

cl_mem out_tensor = clCreateBufferWithProperties(
  ctx, {CL_MEM_TENSOR_EXP, out_desc, 0},
  CL_MEM_USE_HOST_PTR | CL_MEM_WRITE_ONLY, 0, out_data.data(), &err);

// Create a kernel that operates on the tensors and is possibly
// optimized for them using via yet realized API extension.
cl_kernel batched_matmul_kernel = create_batched_matmul_kernel(
  ctx, device_span, in1_desc, in2_desc, out_desc);

clSetKernelArg(batched_matmul_kernel, 0, sizeof(cl_mem), &in0_tensor);
clSetKernelArg(batched_matmul_kernel, 1, sizeof(cl_mem), &in1_tensor);
clSetKernelArg(batched_matmul_kernel, 2, sizeof(cl_mem), &out_tensor);

// Required command for transferring data to layout-opaque tensors and

> // from it to elsewhere.

Reviewer: ```suggestion
// from it elsewhere.
```

clEnqueueExportToTensor(
  cmd_q, in0_tensor, false, {0, 0, 0}, {0, 0, 0}, {b, m, k},
  nullptr, nullptr, in0_data.data(), 0, nullptr, nullptr);

clEnqueueNDRangeKernel(
  cmd_q, batched_matmul_kernel, 3, matmul_grid, nullptr, nullptr, 0, nullptr, nullptr);

clEnqueueMapBuffer(
  cmd_q, out_tensor, CL_TRUE, CL_MAP_READ, 0, b * m * n, 0, nullptr, nullptr);
----

> == Issues

Reviewer: ```suggestion
== Issues and Open Questions
```

> . Should we have support for tensors with undefined shape and tensors

Reviewer: ```suggestion
. Should we support tensors with undefined shape and tensors
```

Reviewer: Should we also look at Khronos NNEF or is it inactive now?

  with unknown / symbolic dimension sizes like in ONNX?
+
--
// https://onnx.ai/onnx/repo-docs/ShapeInference.html
*UNRESOLVED*
--

. Should we define OpenCL C language features for accessing tensors?
+
--

> *RESOLVED*: OpenCL C support for tensors can be introduced later in a

Reviewer: I'd just add those built-ins to this extension right away to avoid expanding the extension fragmentation.

           separate extension. Built-in kernels may benefit from this
           extension as it is.
--

. What is the use case of `cl_tensor_layout_blas_pitch_exp`?
+
--
*UNRESOLVED*
--

. Should image types be extended instead of adding a separate tensor type?
+
--

> *UNRESOLVED*

Reviewer: I think it logically might be the other way around: images a special case of tensors. Perhaps something to consider in the longer run if this one takes off.

--

== Version History

[cols="5,15,15,70"]
[grid="rows"]
[options="header"]
|====
| Version | Date       | Author          | Changes
| 0.1.0   | 2023-11-23 | Henry Linjamäki | *Initial revision*

| 0.2.0   | 2024-8-14 | Henry Linjamäki |

* Rework document structure match to the cl_khr_extension_template.

* Added clEnqueueCopyTensor.

> * Added API for setting memory layout for tensors.

Reviewer: ```suggestion
* Added an API for setting the memory layout for tensors.
```

  rank is greater than two.

// ^ This condition is meant to ensure that the tensor elements at different
// coordinates don't alias in memory.

* _ml_type_ defines the memory layout via enumerators which corresponds to
  predefined configurations of `cl_tensor_layout_blas_exp` structure
  as listed in <<tensor-layout-ml-types,ML tensor layout type>> table.

The memory layout descriptions map tensor coordinates to buffer's
memory byte locations respect to buffer's base address as in the

> followed in pseudo C code example:

Reviewer: ```suggestion
following pseudo C code example:
```
